### PR TITLE
Don't delete branch if other PRs with this branch are open

### LINKED ISSUE #18149: Don't view delete branch button if other PRs with this branch are open
### Feature Description

If I merge a PR, the delete branch is also visible if there is another open PR from this branch. It'd probably good to hide the button if there are other open PRs.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by a1012112796**
fix #18149

### CODE REVIEW

File: routers/api/v1/repo/pull.go [hunk 1]
```diff
@@ -873,6 +873,17 @@ func MergePullRequest(ctx *context.APIContext) {
 	log.Trace("Pull request merged: %d", pr.ID)
 
 	if form.DeleteBranchAfterMerge {
+		// Don't cleanup when other pr use this branch as head branch
+		prs, err := models.GetUnmergedPullRequestsByHeadInfo(pr.HeadRepoID, pr.HeadBranch)
```
Comment (lunny) [inline routers/api/v1/repo/pull.go]: How about just counting it but not load them?

**Review by lunny — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

File: models/pull_list.go [hunk 2]
```diff
@@ -59,6 +59,16 @@ func GetUnmergedPullRequestsByHeadInfo(repoID int64, branch string) ([]*PullRequ
 		Find(&prs)
 }
 
+// CheckUnmergedPullRequestsByHeadInfo check if the pull requests that are open and has not been merged
+// by given head information (repo and branch) exist.
+func CheckUnmergedPullRequestsByHeadInfo(repoID int64, branch string) (bool, error) {
```
Comment (Gusted) [inline models/pull_list.go]:
```suggestion
// HasUnmergedPullRequestsByHeadInfo checks if there are open and not merged pull request
// by given head information (repo and branch)
func HasUnmergedPullRequestsByHeadInfo(repoID int64, branch string) (bool, error) {
```

File: models/pull_test.go [hunk 3]
```diff
@@ -107,6 +107,18 @@ func TestGetUnmergedPullRequest(t *testing.T) {
...
 }
 
+func TestCheckUnmergedPullRequestsByHeadInfo(t *testing.T) {
+	assert.NoError(t, unittest.PrepareTestDatabase())
+
+	exist, err := CheckUnmergedPullRequestsByHeadInfo(1, "branch2")
+	assert.NoError(t, err)
+	assert.Equal(t, true, exist)
+
+	exist, err = CheckUnmergedPullRequestsByHeadInfo(1, "not_exist_branch")
```
Comment (Gusted) [inline models/pull_test.go]:
```suggestion
func TestHasUnmergedPullRequestsByHeadInfo(t *testing.T) {
	assert.NoError(t, unittest.PrepareTestDatabase())

	exist, err := HasUnmergedPullRequestsByHeadInfo(1, "branch2")
	assert.NoError(t, err)
	assert.Equal(t, true, exist)

	exist, err = HasUnmergedPullRequestsByHeadInfo(1, "not_exist_branch")
```

File: routers/api/v1/repo/pull.go [hunk 4]
```diff
@@ -873,6 +873,17 @@ func MergePullRequest(ctx *context.APIContext) {
 	log.Trace("Pull request merged: %d", pr.ID)
 
 	if form.DeleteBranchAfterMerge {
+		// Don't cleanup when other pr use this branch as head branch
+		exist, err := models.CheckUnmergedPullRequestsByHeadInfo(pr.HeadRepoID, pr.HeadBranch)
```
Comment (Gusted) [inline routers/api/v1/repo/pull.go]:
```suggestion
		exist, err := models.HasUnmergedPullRequestsByHeadInfo(pr.HeadRepoID, pr.HeadBranch)
```

File: routers/api/v1/repo/pull.go [hunk 5]
```diff
@@ -873,6 +873,17 @@ func MergePullRequest(ctx *context.APIContext) {
...
+		if err != nil {
+			ctx.ServerError("CheckUnmergedPullRequestsByHeadInfo", err)
```
Comment (Gusted) [inline routers/api/v1/repo/pull.go]:
```suggestion
			ctx.ServerError("HasUnmergedPullRequestsByHeadInfo", err)
```

File: routers/api/v1/repo/pull.go [hunk 6]
```diff
@@ -873,6 +873,17 @@ func MergePullRequest(ctx *context.APIContext) {
 	log.Trace("Pull request merged: %d", pr.ID)
 
 	if form.DeleteBranchAfterMerge {
+		// Don't cleanup when other pr use this branch as head branch
```
Comment (Gusted) [inline routers/api/v1/repo/pull.go]:
```suggestion
		// Don't cleanup when there are other PR's that use this branch as head branch.
```

File: routers/web/repo/issue.go [hunk 7]
```diff
@@ -1600,11 +1600,23 @@ func ViewIssue(ctx *context.Context) {
...
+
+		isPullBranchDeletable := canDelete &&
 			pull.HeadRepo != nil &&
 			git.IsBranchExist(ctx, pull.HeadRepo.RepoPath(), pull.HeadBranch) &&
 			(!pull.HasMerged || ctx.Data["HeadBranchCommitID"] == ctx.Data["PullHeadCommitID"])
 
+		if isPullBranchDeletable && pull.HasMerged {
+			exist, err := models.CheckUnmergedPullRequestsByHeadInfo(pull.HeadRepoID, pull.HeadBranch)
+			if err != nil {
+				ctx.ServerError("CheckUnmergedPullRequestsByHeadInfo", err)
```
Comment (Gusted) [inline routers/web/repo/issue.go]:
```suggestion
			exist, err := models.HasUnmergedPullRequestsByHeadInfo(pull.HeadRepoID, pull.HeadBranch)
			if err != nil {
				ctx.ServerError("HasUnmergedPullRequestsByHeadInfo", err)
```

File: routers/web/repo/pull.go [hunk 8]
```diff
@@ -1051,6 +1051,17 @@ func MergePullRequest(ctx *context.Context) {
 	log.Trace("Pull request merged: %d", pr.ID)
 
 	if form.DeleteBranchAfterMerge {
+		// Don't cleanup when other pr use this branch as head branch
+		exist, err := models.CheckUnmergedPullRequestsByHeadInfo(pr.HeadRepoID, pr.HeadBranch)
+		if err != nil {
+			ctx.ServerError("CheckUnmergedPullRequestsByHeadInfo", err)
```
Comment (Gusted) [inline routers/web/repo/pull.go]:
```suggestion
		exist, err := models.HasUnmergedPullRequestsByHeadInfo(pr.HeadRepoID, pr.HeadBranch)
		if err != nil {
			ctx.ServerError("HasUnmergedPullRequestsByHeadInfo", err)
```

File: routers/web/repo/pull.go [hunk 9]
```diff
@@ -1222,6 +1233,17 @@ func CleanUpPullRequest(ctx *context.Context) {
 		return
 	}
 
+	// Don't cleanup when other pr use this branch as head branch
+	exist, err := models.CheckUnmergedPullRequestsByHeadInfo(pr.HeadRepoID, pr.HeadBranch)
+	if err != nil {
+		ctx.ServerError("CheckUnmergedPullRequestsByHeadInfo", err)
```
Comment (Gusted) [inline routers/web/repo/pull.go]:
```suggestion
	exist, err := models.HasUnmergedPullRequestsByHeadInfo(pr.HeadRepoID, pr.HeadBranch)
	if err != nil {
		ctx.ServerError("HasUnmergedPullRequestsByHeadInfo", err)
```

File: routers/web/repo/pull.go [hunk 10]
```diff
@@ -1222,6 +1233,17 @@ func CleanUpPullRequest(ctx *context.Context) {
 		return
 	}
 
+	// Don't cleanup when other pr use this branch as head branch
```
Comment (Gusted) [inline routers/web/repo/pull.go]:
```suggestion
	// Don't cleanup when there are other PR's that use this branch as head branch.
```

**Review by Gusted — COMMENTED** [review]
Just a _few_ nitpicks 😅

**Review by Gusted — APPROVED** [review]

**Pull request merged**